src/Utilities/CssDocument.php

```diff
@@ -0,0 +1,283 @@
+     *
+     * @param string $css CSS with comments, `@import` and `@charset` removed
+     *
+     * @return array{0: string, 1: string}
+     *         The first element is the CSS with the at-rules removed.  The second element contains a concatenation of
+     *         the valid at-rules, each followed by whatever whitespace followed it in the original CSS (so that either
+     *         unminified or minified formatting is preserved); if there were no at-rules, it will be an empty string.
+     */
+    private function extractNonConditionalAtRules(string $css): array
+    {
+        $possiblyModifiedCss = $css;
+        $atRules = [];
+
+        while (\preg_match(self::NON_CONDITIONAL_AT_RULE_PATTERN, $possiblyModifiedCss, $matches)) {
+            [$fullMatch, $atRuleName] = $matches;
+
+            if ($this->isValidAtRule($atRuleName, $fullMatch)) {
+                $atRules[] = $fullMatch;
+            }
+
+            $possiblyModifiedCss = \str_replace($fullMatch, '', $possiblyModifiedCss);
+        }
+
+        return [$possiblyModifiedCss, \implode('', $atRules)];
+    }
+
+    /**
+     * Tests if an at-rule is valid.  Currently only `@font-face` rules are checked for validity; others are assumed to
+     * be valid.
+     *
+     * @param string $atIdentifier name of the at-rule with the preceding at sign
+     * @param string $rule full content of the rule, including the at-identifier
+     *
+     * @return bool
+     */
+    private function isValidAtRule(string $atIdentifier, string $rule): bool
+    {
+        if (\strcasecmp($atIdentifier, '@font-face') === 0) {
+            return \stripos($rule, 'font-family') !== false && \stripos($rule, 'src') !== false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Splits input CSS code into an array of parts for different media queries, in order.
+     * Each part is an array where:
+     *
+     * - key "css" will contain clean CSS code (for @media rules this will be the group rule body within "{...}")
+     * - key "media" will contain "@media " followed by the media query list, for all allowed media queries,
+     *   or an empty string for CSS not within a media query
+     *
+     * Example:
+     *
+     * The CSS code
+     *
+     *   "@import "file.css"; h1 { color:red; } @media { h1 {}} @media tv { h1 {}}"
+     *
+     * will be parsed into the following array:
+     *
+     *   0 => [
+     *     "css" => "h1 { color:red; }",
+     *     "media" => ""
+     *   ],
+     *   1 => [
+     *     "css" => " h1 {}",
+     *     "media" => "@media "
+     *   ]
+     *
+     * @param array<array-key, string> $allowedMediaTypes
+     *
+     * @return array<int, array{css: string, media: string}>
+     */
+    private function splitCssAndMediaQuery(array $allowedMediaTypes): array
+    {
+        $mediaTypesExpression = '';
+        if (!empty($allowedMediaTypes)) {
+            $mediaTypesExpression = '|' . \implode('|', $allowedMediaTypes);
```

Comment:
We need to use `preg_quote` here (with the `$delimiter` parameter).

Reply:
That latent bug has been around for a long time, though unlikely to arise in practice as media types are pretty much restricted to alphanumeric characters (I think).  Fixed.